crates/red_knot/src/lint.rs

```diff
@@ -140,6 +142,66 @@ fn lint_unresolved_imports(context: &SemanticLintContext) -> QueryResult<()> {
     Ok(())
 }
 
+fn lint_bad_overrides(context: &SemanticLintContext) -> QueryResult<()> {
+    // TODO possibly we should have a special marker on the real typing module (from typeshed) so
+    // if you have your own "typing" module in your project, we don't consider it THE typing module
+    let Some(typing_module) = context.db.resolve_module(ModuleName::new("typing"))? else {
+        debug_assert!(false, "typing module should always exist");
+        return Ok(());
+    };
+    let typing_file = context.db.module_to_file(typing_module)?;
+    let typing_table = context.db.symbol_table(typing_file)?;
+    let Some(typing_override) = typing_table.root_symbol_id_by_name("override") else {
+        debug_assert!(false, "typing.override should exist");
+        return Ok(());
+    };
+
+    // TODO we should maybe index definitions by type instead of iterating all, or else iterate all
+    // just once, match, and branch to all lint rules that care about a type of definition
+    for (symbol, definition) in context.symbols().all_definitions() {
+        if !matches!(definition, Definition::FunctionDef(_)) {
+            continue;
+        }
+        let ty = context
+            .db
+            .infer_definition_type(context.file_id, symbol, definition.clone())?;
+        let Type::Function(func) = ty else {
+            debug_assert!(false, "type of a FunctionDef should always be a Function");
+            continue;
+        };
+        let Some(class) = func.get_containing_class(context.db)? else {
+            // not a method of a class
+            continue;
+        };
+        let mut override_decorated = false;
+        for deco_ty in func.function(context.db)?.decorators() {
+            let Type::Function(deco_func) = deco_ty else {
+                continue;
+            };
+            if deco_func.file() == typing_file && deco_func.symbol(context.db)? == typing_override {
+                override_decorated = true;
+                break;
+            }
+        }
```

Comment:
This is slightly more verbose but feels slightly cleaner to me -- but it's just a stylistic preference, feel free to ignore

```suggestion
        let override_decorated = func
            .function(context.db)?
            .decorators()
            .iter()
            .filter_map(|deco_type| {
                if let Type::Function(deco_func) = deco_type {
                    Some(deco_func)
                } else {
                    None
                }
            })
            .any(|deco_func| {
                deco_func.file() == typing_file
                    && deco_func
                        .symbol(context.db)
                        .is_ok_and(|symbol| symbol == typing_override)
            });
```

Reply:
I played around with options like this for a while (including this very one), but I don't think it's good to just ignore an error result from the `.symbol(context.db)` call like this. The current code bubbles it up to the caller.

Reply:
That makes sense. You could `.collect()` them into a single `Result<Vec<_>>` like this, and then bubble up the error to the caller:

```suggestion
        let override_decorated = func
            .function(context.db)?
            .decorators()
            .iter()
            .filter_map(|deco_type| {
                let Type::Function(deco_func) = deco_type else {
                    return None;
                };
                if deco_func.file() != typing_file {
                    return None;
                }
                Some(deco_func.symbol(context.db))
            })
            .collect::<QueryResult<Vec<SymbolId>>>()?
            .iter()
            .any(|symbol_id| *symbol_id == typing_override);
```

Reply:
Maybe something like this would be best -- it's pretty similar to your current code, but the `override_decorated` variable doesn't have to be marked as mutable in the main `for` loop, which I think is the main thing I didn't really like about your current code

```suggestion
        let override_decorated = {
            let mut found_override_decorator = false;
            for deco_ty in func.function(context.db)?.decorators() {
                let Type::Function(deco_func) = deco_ty else {
                    continue;
                };
                if deco_func.file() == typing_file
                    && deco_func.symbol(context.db)? == typing_override
                {
                    found_override_decorator = true;
                    break;
                }
            }
            found_override_decorator
        };
```

Reply:
Yeah, I'd also considered the collect-to-a-Vec option, but allocating unnecessarily didn't seem worth it. The `try_fold` option is nice! And yeah I agree this should probably just be a `has_decorator` function.